File: src/main/java/org/stt/query/WorkTimeQueries.java
package org.stt.query;

import org.stt.reporting.WorkingtimeItemProvider;
import org.stt.text.ItemCategorizer;
import org.stt.time.DateTimes;
import org.stt.time.Interval;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.time.DayOfWeek;
import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Objects;

@Singleton
public class WorkTimeQueries {
    private WorkingtimeItemProvider workingtimeItemProvider;
    private ItemCategorizer itemCategorizer;
    private TimeTrackingItemQueries timeTrackingItemQueries;

    @Inject
    public WorkTimeQueries(WorkingtimeItemProvider workingtimeItemProvider, ItemCategorizer itemCategorizer,
                           TimeTrackingItemQueries timeTrackingItemQueries) {
        this.workingtimeItemProvider = Objects.requireNonNull(workingtimeItemProvider);
        this.itemCategorizer = Objects.requireNonNull(itemCategorizer);
        this.timeTrackingItemQueries = Objects.requireNonNull(timeTrackingItemQueries);
    }

    public Duration queryRemainingWorktimeToday() {
        LocalDateTime now = DateTimes.preciseToSecond(LocalDateTime.now());
        LocalDate today = now.toLocalDate();
        Duration workedTime = queryWorktime(Interval.ofDay(today).withEnd(now));
        return workingtimeItemProvider.getWorkingTimeFor(today).getMin().minus(workedTime);
    }

    public Duration queryWeekWorktime() {
        LocalDateTime now = DateTimes.preciseToSecond(LocalDateTime.now());
        LocalDate monday = now.toLocalDate().with(DayOfWeek.MONDAY);
        Interval currentWeek = Interval.between(monday.atStartOfDay(), now);
        return queryWorktime(currentWeek);
    }

    public Duration queryWorktime(Interval interval) {
        Criteria criteria = new Criteria();
        criteria.withStartNotBefore(interval.getStart());
        criteria.withStartBefore(interval.getEnd());
        return timeTrackingItemQueries.queryItems(criteria)
                .filter(item -> itemCategorizer.getCategory(item.getActivity()) == ItemCategorizer.ItemCategory.WORKTIME)
                .map(item -> {
                    LocalDateTime end = item.getEnd().orElse(interval.getEnd());
                    return Duration.between(item.getStart(), end);
                })
                .reduce(Duration.ZERO, Duration::plus);
    }
}

File: src/main/kotlin/org/stt/query/WorkTimeQueries.kt
package org.stt.query

import org.stt.reporting.WorkingtimeItemProvider
import org.stt.text.ItemCategorizer
import org.stt.time.Interval
import org.stt.time.asInterval
import org.stt.time.preciseToSecond
import org.stt.time.until
import java.time.DayOfWeek
import java.time.Duration
import java.time.LocalDateTime
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WorkTimeQueries @Inject
constructor(private val workingtimeItemProvider: WorkingtimeItemProvider,
            private val itemCategorizer: ItemCategorizer,
            private val timeTrackingItemQueries: TimeTrackingItemQueries) {

    fun queryRemainingWorktimeToday(): Duration {
        val now = LocalDateTime.now().preciseToSecond()
        val today = now.toLocalDate()
        val workedTime = queryWorktime(today.asInterval().withEnd(now))
        return workingtimeItemProvider.getWorkingTimeFor(today).min.minus(workedTime)
    }

    fun queryWeekWorktime(): Duration {
        val now = LocalDateTime.now().preciseToSecond()
        val monday = now.toLocalDate().with(DayOfWeek.MONDAY)
        val currentWeek = monday.atStartOfDay() until now
        return queryWorktime(currentWeek)
    }

    fun queryWorktime(interval: Interval): Duration {
        val criteria = Criteria()
        criteria.withStartNotBefore(interval.start)
        criteria.withStartBefore(interval.end)
        return timeTrackingItemQueries.queryItems(criteria)
                .filter { (activity) -> itemCategorizer.getCategory(activity) == ItemCategorizer.ItemCategory.WORKTIME }
                .map { (_, start, end) ->
                    Duration.between(start, end ?: interval.end)
                }
                .reduce(Duration.ZERO, { obj, duration -> obj.plus(duration) })
    }
}
